datalake: add scratch space usage limit

dotnwat opened the pull request:
This PR adds a disk space usage monitor in the datalake manager that is intended apply a soft limit to the amount of staging data generated on-disk by datalake. It is modeled after space management in that it (1) periodically calculates the current usage and (2) responds to overages by dispatching to each shard that certain resources be promptly released. In this case the resources are the flushed bytes that translators may be holding on to. Requests are satisfied by stopping translation, uploading the current flushed data, and removing that data from disk.

The primary components are:

### Datalake manager

* Simple datalake_manager control loop
* Calculates total usage by querying all translators for the current flushed bytes count
* Notably this avoids stat() system calls because the statistics are maintained in memory
* Selects the top disk usage translators necessary to meet the target disk usage size
* Dispatches requests to each translator selected to "finish" promptly

### Translator support for fast "finish"

The mechanism for a translator to finish early existed before this work in the form of preemption due to memory overage. When the preemption occurred the translator would stop, flush its state to disk, and then upload+delete the staged data.

This strategy works well for memory resources when translators are in a "running" state--stopping them releases the memory resources. However, this doesn't work for disk resources that exist independent of the state of the translator (e.g. waiting or idle). For waiting tasks the solution is relatively simple: put the translator into the running state and arrange for it to immediately execute the "finish" code path.

The solution is not so simple for idle translators that are neither running nor waiting. The complexity arises because the stop_translation interface is not designed to affect state change for an idle translator. In fact, in order for an idle translator to even start running it needs there to be new offsets ready for translation. That isn't good enough for disk space management because an idle partition may take up a lot of disk space and never be runnable.

To solve the problem for idle translators the solution was to add a bit of a "ugly" interface for this specific use case that to a certain extent depends on knowledge of the underlying partition translator implementation. The scheduler effectively sets a flag on the translator which is checked in its "new offsets" polling loop.

This narrow, "ugly" interface can certainly be eliminated by expanding the expressiveness of the existing translator interface to accommodate control over idle translators, but that may be a bit too much this late in the game.

TODO:
* been testing manually, will add ducktape tests next

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

dotnwat committed:
test: add metric expectation assertion

Makes the query more robust by not allowing the user to unexpectedly
receive a zero for something like a typo.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
datalake: run a global disk space usage monitor

Modeled after space management, datalake manager monitors the usage of
the datalake scratch space directory. When size exceeds a configured
threshold the monitor makes requests to the scheduler to finish
translators and upload data.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
datalake: add translator finish_translation interface

Adds an interface to request a translator to finish translation even
when the translator is idle (neither running nor waiting). This is
important for making a translator responsive in low disk usage
situations since meaningful amounts of on-disk data may be under
management of a translator and stop_translation--the only current
preemption mechanism--is designed to stop translations that are in the
running state.

Ideal finish_translation is merged with a more generic version of
stop_translation, but that is a bigger project (in the context of the
partition_translator implementation) because the life cycle of
start/stop translation interfaces is tightly coupled with the scheduler
(e.g. intrusive list membership, expecte statuses and notification of
state changes).

More details:

The partition translator operates in a nested structure. At the outer
layer is a polling loop that waits for new offsets to translate:

   translate_until_stopped
      poll(fetch_translation_offsets)

once new offsets are available then the next layer waits for the
scheduler to allow it to run and process the offsets:

   translate_until_stopped
      poll(fetch_translation_offsets)
         run_one_translation_iteration

Notice that there are two places where control can block: at the outer
later if there are no new offsets to process (idle topic) and in the
nested level waiting for the scheduler to schedule the translation.
Currently the translation interface stop_translation is only capable of
preempting a running translation. That is, one which is executing at the
nested level.

This is problematic when it comes to controlling the disk space usage of
a translator because we need to take into account the usage of idle
translators which are neither running or waiting to be schedule. They
may have a lot of staging data, and need to run shortly to finish up the
current translation. In this case they are likely polling in the outer
layer waiting for offsets. However, currently the only way to preempt
this outer layer polling is by completely shutting down the translator
using an abort source. That is not what we want to do since we'd have to
rebuild the translator.

What this commit does is add a new translator interface,
finish_translation which signals to the translator to finish, and can be
used when the translator is in any state. The partition translator
implements the interface by setting a flag which the outer polling layer
examines in order to break out of its polling loop even if there are no
new offsets to translate.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat commented:
/ci-repeat 1

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c4da94be-29aa-447f-8765-c8007b72d93f-->
<!--TEST RESULTS-->
<!--8c043417-06d4-4477-9bff-d9c20df01c0a-->
<!--TEST RESULTS-->
<!--9069e9b5-d59e-4e86-8bd4-871ff7db2cf0-->
<!--TEST RESULTS-->
<!--5a0699e0-d520-43c8-84a1-9143f4eed84a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63942">build#63942</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/63942#0195e5e2-e01f-44e5-8523-7aef4b1dc47e | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63942#0195e5e2-e01f-44e5-8523-7aef4b1dc47e | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63942#0195e5dd-8e46-46eb-b594-24d4b93ab322 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63942#0195e5dd-8e46-4573-ba8d-baf51803e144 | FLAKY | 1/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_spec_evolution_rpcn.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63942#0195e5dd-8e46-46eb-b594-24d4b93ab322 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/63942#0195e5e2-e020-4f0c-a688-4c9f729b1bce | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63942#0195e5e2-e021-463a-9d4c-7be56cc98a2f | FLAKY | 1/2 |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=True.allow_gaps_cluster_level=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63942#0195e5e2-e022-4080-8bcf-039439e08eb5 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63957">build#63957</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=start_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/63957#0195ead5-ffb5-443a-826a-fce86f03e919 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63957#0195ead5-ffb5-443a-826a-fce86f03e919 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63957#0195eac3-013f-413d-963c-f3a8a4b58029 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64005">build#64005</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64005#0195ef8d-1c0d-4b7c-a04a-9061c0aa041b | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64077">build#64077</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.availability_test.AvailabilityTests.test_availability_when_one_node_failed                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64077#0195f441-baf3-4453-843b-cdf5bb53ac07 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.executing.use_alias.True  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64077#0195f441-baf2-4263-8687-9b8d36578655 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64077#0195f441-baf5-49a2-86e6-3d4d9d9e0e89 | FLAKY         | 1/3      |
</details>
<!--/body-->

dotnwat committed:
datalake: add scheduler interface for immediate finish

This commit introduces a new interface:

   scheduler::request_immediate_finish(vector<translator>)

which can be used by an entity external to the scheduler for requesting
that a set of translators be promptly stopped and finished. The primary
use case for this interface are global resource schedulers (e.g. disk
space) which want to inject priority scheduling decisions. Translators
request for immediate finish should promptly upload and remove their
staging data from disk.

This commit adds the basic scheduling component, and the next commit
sprinkles around some changes that make scheduling more responsive.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
datalake: make immediate finish request more responsive

Checks for request in strategic locations and pokes the scheduler when
immediate finish requests are made.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
datalake: silence oom warning when finish is requested

If finish translation has been requested of a running translator then
currently preemption is provided by the stop translation interface which
unfortunatley in its current form implies an out-of-memory condition
which logs a warning. This commit silences the warning when a translator
is preempted and finish translation has also been requested.

Once a generalized, more expressive interface can be used to
finish/stop/preempt a translator some of these flag checks/hacks can be
removed.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
storage: add datalake usage to space management probes

Added to space management probes because (1) we expect that space
management will be used in conjunction with any use of cloud storage
including datalake, and (2) that space management will in short order
take over the role of managing disk usage by datalake.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -520,4 +532,13 @@ void partition_translator::stop_translation() {
       translator_out_of_memory_error{});
 }
+
+void partition_translator::finish_translation() {
```
nit: maybe `set_finish_translation()`? Otherwise sounds like a verb

dotnwat replied:
fixed

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+        /*
+         * state maintained for eviction policy
+         *
+         * elem.first: stable iterator into executor.translators
+         * elem.second: key into translators_for_immediate_finish
+         * elem.third (other) true if on waiting list
```
nit: i think it would be easier to follow below if this were a struct, then we could store the status too and not have to track stopped_status separately

dotnwat replied:
agreed, did this and it cleaned it up a lot!

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+          "Expected a choice to be made from non-empty set");
+
+        // stopped is the key in finish set (if any)
+        std::optional<size_t> stopped;
```
nit: stopped_size?

EDIT: stopped_priority?

dotnwat replied:
this is completely gone in the refactor

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
   executor& executor, const reservations_tracker& mem_tracker) {
+    if (!executor.translators_for_immediate_finish.empty()) {
+        /*
+         * state maintained for eviction policy
```
nit: a brief comment about what we're doing below would be pretty helpful.

dotnwat replied:
added more commentary in the refactor, but this specific chunk of code is gone (e.g. the struct idea you had makes it much clearer)

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+        std::optional<std::tuple<translators::iterator, size_t, bool>>
+          first_other;
+
+        auto& finish = executor.translators_for_immediate_finish;
```
nit: given how tall this code is, IMO it's a bit confusing to refer to this as just "finish" with no units or additional description (e.g. if just looking through the code farther down the page, it's kind of a surprising name to see). Maybe `translators_to_finish` or `translator_to_finish_by_size` or something?

dotnwat replied:
yep, fixed

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+                }
+            }(stopped_status);
+
+            const auto it = finish.find(stopped.value());
```
nit: why did we cache the size instead of the iterator, if we need the iterator here anyway?

dotnwat replied:
the use of the struct makes it clear (e.g. now it is called `translators_to_finish_key`. the reason you can't hold on to an iterator is because it may be invalidated while looping through the choices when we erase translators that don't exist in the global set tracked by the exector.

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+        }
+
+        /*
+         * TODO: currently we schedule one immediate finish task at a time, but
```
Maybe a dumb question, but why aren't we doing more? It seems surprising to proceed after just one, and looks (maybe deceptively?) easy?

Also looking at the space manager, it logs that it will try to reclaim some amount of space. It will almost certainly be construed as a bug when it only ends up freeing up one translator's worth of disk space

dotnwat replied:
> It will almost certainly be construed as a bug when it only ends up freeing up one translator's worth of disk space

the choice of doing "one" may be misleading. we aren't doing "one", we just aren't doing more than one in parallel. when we are done handling one, control will return back to `_scheduling_policy->schedule_one_translation(` and `scheduler::main` where execution will return to `on_resource_exhausion` (because there are still pending requests for immediate finish) and another choice will be made.

were you instead talking about doing them in parallel? if so, then yeh we could do more in parallel and probably should add a knob for this.

dotnwat replied:
offline discussion here:

* we'll finish all the translators in on_resource_exhaustion because that's a more accurate representation of what its intended purpose is as a method
* we're going ot add a config value that allows more than one to finish in parallel

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+                /*
+                 * Unlike the case for running translators (and those that are
+                 * waiting, but put into the running state), idle translators
+                 * are not on a list to count to ascertain when they have
+                 * completed finishing. for idle translators we poll their
+                 * finishing state request which we made above.
```
Can you also mention if/how we're guaranteed that the idle translator is expected to be scheduled soon?

dotnwat replied:
its described in the next sentence that the translator is in a polling state when idle. but i rewrote it to make it clearer it was kind of vague.

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+              tid);
+
+            if (stopped_status == status::idle) {
+                /*
+                 * Unlike the case for running translators (and those that are
+                 * waiting, but put into the running state), idle translators
+                 * are not on a list to count to ascertain when they have
+                 * completed finishing. for idle translators we poll their
+                 * finishing state request which we made above.
```
Hmm it's true that we can't do the running check for idle translators, but could we do _this_ if block for all translators? The end condition seems like it'd work for the others, but I'm not sure. It'd be nice though because then everything just takes the same code path

dotnwat replied:
yes i think it will work for all the states. nice catch

andrwng commented on src/v/datalake/translation/scheduling.h:
```diff
@@ -307,6 +316,7 @@ struct executor {
       running;
     intrusive_list<translator_executable, &translator_executable::_waiting_hook>
       waiting;
+    absl::btree_map<size_t, translator_id> translators_for_immediate_finish;
```
nit: can we use a named "priority" type here or somesuch? This looks like it is keyed by size

dotnwat replied:
done

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+          first_other;
+
+        auto& finish = executor.translators_for_immediate_finish;
+        for (auto it = finish.begin(); it != finish.end();) {
+            // map translator_id -> translator_executable
```
nit: maybe we can avoid tracking both choices as

```
choice = None
for prio, it in finish:
  if it not in translators:
    finish.erase(it)
    continue
  if translators.find(it) is running:
    choice = it
    break;

if choice is None:
  choice = finish.begin()
```

It's nice to only have to keep track of one thing at a time (first look for one that's running, otherwise we know which one is high priority)

Plus, I think the `if (first_running) else if (first_other)` block can collapse into a switch on the status of `choice`

dotnwat replied:
using the struct to avoid tracking two choices (https://github.com/redpanda-data/redpanda/pull/25618#discussion_r2021967784) cleaned this up a lot. selecting the backup (non-running) choice after breaking out of the loop, though, means we have to search the the global translator index again, so i left that part in the loop.

dotnwat replied:
fixed it. along with https://github.com/redpanda-data/redpanda/pull/25618#discussion_r2021967784 it cleaned up nicely. it has an extra index loopkup, but the code clarity is probably worht it.

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +289,192 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+            ++it;
+        }
+
+        vassert(
+          first_running.has_value() || first_other.has_value()
+            || finish.empty(),
+          "Expected a choice to be made from non-empty set");
```
What guarantees that we didn't erase them all?

dotnwat replied:
nothing. but if we did erase them all then first_xxx wouldn't have a value, finish.empty() would be true, and then we'd fall through the rest of the conditions below this. did you spot a logic flaw?

andrwng replied:
Ah 🤦 just entirely overlooked finish.empty()!

andrwng commented on src/v/datalake/translation/scheduling.cc:
```diff
@@ -428,9 +434,12 @@ ss::future<> scheduler::main() {
+          _mem_tracker->memory_exhausted(),
+          _executor.translators_for_immediate_finish.size());
+        if (
+          _mem_tracker->memory_exhausted()
+          || !_executor.translators_for_immediate_finish.empty()) {
             co_await _scheduling_policy->on_resource_exhaustion(
```
I'm squinting a bit trying to convince myself this call to on_resource_exhaustion never blocks indefinitely, waiting for a schedule_one_translation() call that isn't coming... Any reason for concern there?

dotnwat replied:
yeh, it's tightly coupled which isn't great. but it looks safe to me. we set the `finish bit` once, and then the translator will always reset it after a few seconds in its idle polling loop (see the `ss::defer` that clears it), or, if it is running/waiting it will receive the stop signal which brings it back to the same loop containing the same RAII device which clears the bit.

dotnwat replied:
>on_resource_exhaustion never blocks indefinitely, waiting for a schedule_one_translation() call that isn't coming

these run synchronously in the same fiber. are you asking if `on_resource_exhaustion` could block indefinitely and then `schedule_one_translation` would never run?

dotnwat replied:
from slackadoo

andrew

> like if all trnaslators are idle but take up space, we get to on_resource_exhaustion and set that the translator should finish, but the scheduler isn’t given the opportunity to kick it? (assuming schedule_one_translation is what finishes it, i forget)

noah

> idle translators are not scheduled into the running (or waiting and then running) state to finish because they are blocked polling for new offsets to translate. instead, that same polling loop notices the request to finish via set_finish_translation and skip all of the offset querying of the underlying partition, building a reader, etc... and goes directly to jail

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -411,7 +419,9 @@ ss::future<> partition_translator::translate_until_stopped() {
+        // run_one_translation_iteration which uses the flag to silence the
+        // out-of-memory warning message.
         auto clear_finish_request = ss::defer(
           [this] { _finish_translation_requested = false; });
```
If we just made stop_translator() accept an exception, and used some new translation_stop_requested_exception, could we avoid the additional orchestration of this _finish_translation_requested? It seems a bit more explicit and easier to follow IMO, though maybe it misses out on something that the current impl handles?

dotnwat replied:
yes i think so. the catch is that we end up having to change the semantics of stop_translation such that it does one thing for running translator and another thing for a idle translator. that's not a problem per se, i just didn't want to rock the boat too much. happy to prototype that tho it should be quick

andrwng replied:
Ahh got it, makes sense. In that case, I don't mind it as is (though definitely open to the change if it does end up seeming more straightforward)

dotnwat committed:
datalake: add test for space control of idle translators

Tests that idle translators can have their data on disk finished in low
disk situations when other conditions like timeout and data flush limits
do not apply.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
datalake: make configs not require restart

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

andrwng reviewed (commented):
Mostly a bunch of suggestions to make the code a bit less branchy, but also had a few questions

dotnwat commented:
Force push: fixes no-restart flag on configuration options, adds first ducktape test

dotnwat commented:
Force push: test should only run in release mode

dotnwat commented:
Forced push https://github.com/redpanda-data/redpanda/compare/58807a6b3189d52625f0468a32cab53519ff6ded..f9b9f88d8aa4fb6c85caed26b0352603b9e8c857 addresses all resolved feedback, i think!

nvartolomei commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -289,8 +290,77 @@ ss::future<> fair_scheduling_policy::schedule_one_translation(
+
+        co_await ss::max_concurrent_for_each(
+          choices, max_parallel, [this, &executor](auto& choice) {
+              return finish_translator(executor, choice);
```
🤔 I'm try to rely on finish translator not being called in the scheduler context in https://github.com/redpanda-data/redpanda/pull/25674/commits/d04518ca17387e6c0d8ad2ce193fed0c4559fb43 (part of https://github.com/redpanda-data/redpanda/pull/25674) as it can take really long in some cases

i'm a bit more concerned by the fact this blocks memory exhaustion preemption possible for many minutes in the case there are translators for which we want to call finish?

nvartolomei replied:
Alternatively: we should only signal that finish should called and until space is freed no additional translation should take place.

Btw, does the current implementation achieve that at all? Or if finish is taking too long (uploading) then we'll just schedule additional translations and continue growing the disk space without any real bound?

dotnwat replied:
it's a good point. i thought about this too, but i don't think there is a great answer.

if we are over the memory budget, then we really should be mindful of improving the situation. but improving the situation will mean flushing data to disk (at least in the current implementation), which might make disk usage worse if we are also in low disk usage situation :)

i think ideally if we were in both states at the same time, then it would be better to throw away the in-memory state of running translators without flushing, and repeat that work when resources became available again. i don't know if we have that capability right now, though. let's assume we don't.

so which do we prioritize? let's also assume we aren't going to do any "fancy" accounting of resources or consider in the code about hypothetical decisions.

i think you can mostly have your cake and eat it too: in the latest version of this code we run multiple "finish" requests in parallel. i think if we have both we can fold them together and run "some of both" classes at the same time. thoughts?

andrwng commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -317,4 +397,129 @@ ss::future<> fair_scheduling_policy::on_resource_exhaustion(
+
+ss::future<> fair_scheduling_policy::finish_translator(
+  executor& executor, finish_choice_info choice) {
+    auto& executable = choice.it->second;
+    executable.translator_ptr()->set_finish_translation();
```
What guarantees us that the iterator is still valid when we get here (e.g. what if the partition moved)? Should we just store the translator id in `choice` and do the lookup again? Same question for down below when we're polling

dotnwat replied:
yep, you're right. it was guaranteed earlier when we were doing one-at-a-time finishing because there were no scheduling points, but that doesn't hold with running them in parallel.

fixed

dotnwat replied:
> below when we're polling

down there (if are talking about the same location) we do look-up the translator again for precisely the same reason you highlight. are we talking about the same location and the same situation?

andrwng replied:
Ah you're right, I was thinking `it` was still `choice.it`, sorry!

dotnwat replied:
i replaced `choice.it` with `choice.id` and just repeat the lookups when necessary in case the translators peace out unexpectedly

dotnwat commented:
Pushed to fix 🐍 formatting

andrwng reviewed (approved):
Would be nice to squash the last two commits into the history so we don't have buggy code in our history, if we can avoid it

dotnwat commented:
> Would be nice to squash the last two commits into the history so we don't have buggy code in our history, if we can avoid it

@andrwng yeh, thanks its a bad habit. fixed. see force push https://github.com/redpanda-data/redpanda/compare/6046cf651e19aac703a1e10267e39f9b715390d0..04b936404feeb04dc094b048f9a021db928bdad4 with zero size delta.

andrwng reviewed (approved):
Thanks!

WillemKauf commented on src/v/datalake/translation/scheduling.h:
```diff
@@ -307,6 +316,15 @@ struct executor {
+     * finished with lowest value being highest priority.
+     */
+    using finish_priority = size_t;
+    absl::btree_map<finish_priority, translator_id>
```
Do we need a `multimap` here to avoid past mistakes or are the keys guaranteed to be unique? :)

WillemKauf replied:
Ah sorry just read the other bit of code, looks like keys are unique

dotnwat replied:
they are unique (see were the index is populated). good memory!

WillemKauf commented on src/v/config/configuration.cc:
```diff
@@ -4087,6 +4087,40 @@ configuration::configuration()
+      *this,
+      "datalake_disk_usage_overage_coeff",
+      "The datalake disk usage monitor reclaims the overage multiplied by "
+      "this this coefficient to compensate for data that is written during the "
```
> this this coefficient

should be `this coefficient`

dotnwat replied:
will fix thanks

WillemKauf commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -268,6 +275,158 @@ ss::future<> datalake_manager::start() {
+     * thrashing scenarios. this is the same strategy used in space management
+     * to avoid thrashing (see resource_mgm/storage.cc).
+     */
+    const size_t min_size_threshold = 64_MiB;
```
is there a reason this threshold isn't also a tunable property?

dotnwat replied:
maybe it should be. it's really just to prevent some thrashing. i think it is probably low stakes

WillemKauf commented on src/v/datalake/translation/scheduling_policies.cc:
```diff
@@ -317,4 +397,141 @@ ss::future<> fair_scheduling_policy::on_resource_exhaustion(
+            break;
+        }
+
+        if (!choice.has_value()) {
```
This repeated logic makes things harder to parse- Why not just

```cpp
choice = choice_info{.info = curr_info, .finish_key = it->first};

if (curr_info.status == finish_choice_info::status::running) {
   break;
}

// continue looking for a running translator
```

dotnwat replied:
ahh, yeh that is simpler. thanks!

vbotbuildovich commented:
/backport v25.1.x
